src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorExtensions.cs

```diff
@@ -0,0 +1,1754 @@
+            where T : IEquatable<T>, IEqualityOperators<T, T, bool>
+        {
+            if (input.Rank == 1)
+            {
+                return input;
+            }
+            else
+            {
+                T[] values = input.IsPinned ? GC.AllocateArray<T>((int)input._linearLength, input.IsPinned) : (new T[input._linearLength]);
+                nint[] lengths = new nint[input.Rank];
+                Tensor<T> tensor;
+                SpanND<T> ospan;
+                SpanND<T> ispan;
+                nint[] indices;
+                int[] permutation;
+
+                if (axis.IsEmpty)
+                {
+                    lengths = input._lengths.Reverse().ToArray();
+                    permutation = Enumerable.Range(0, input.Rank).Reverse().ToArray();
+                }
+                else
+                {
+                    if (axis.Length != input.Lengths.Length)
+                        throw new ArgumentException("Must provide an axis order for each axis");
+                    for (int i = 0; i < lengths.Length; i++)
+                        lengths[i] = input.Lengths[axis[i]];
+                    permutation = axis.ToArray();
+                }
+                tensor = new Tensor<T>(values, lengths, Array.Empty<nint>(), input._isPinned);
+                nint[] permutedIndices = new nint[tensor.Rank];
+
+                ospan = tensor.AsSpan();
+                ispan = input.AsSpan();
+                indices = new nint[tensor.Rank];
+                for (int i = 0; i < input._linearLength; i++)
+                {
+                    TensorHelpers.PermuteIndices(ref indices, ref permutedIndices, ref permutation);
+                    ospan[permutedIndices] = ispan[indices];
+                    SpanHelpers.AdjustIndices(tensor.Rank - 1, 1, ref indices, input._lengths);
+                }
+
+                return tensor;
+            }
+        }
+        #endregion
+
+        #region TensorPrimitives
+        #region Multiply
+        public static Tensor<T> Multiply<T>(Tensor<T> input, T val)
+            where T : IEquatable<T>, IEqualityOperators<T, T, bool>, IMultiplyOperators<T, T, T>, IMultiplicativeIdentity<T, T>
+        {
+            ReadOnlySpan<T> span = MemoryMarshal.CreateSpan(ref input._values[0], (int)input._linearLength);
+            Tensor<T> output = Create<T>(input.IsPinned, input.Lengths);
+            Span<T> ospan = MemoryMarshal.CreateSpan(ref output._values[0], (int)output._linearLength);
+            TensorPrimitives.Multiply(span, val, ospan);
+            return output;
+        }
+
+        public static Tensor<T> MultiplyInPlace<T>(Tensor<T> input, T val)
+            where T : IEquatable<T>, IEqualityOperators<T, T, bool>, IMultiplyOperators<T, T, T>, IMultiplicativeIdentity<T, T>
+        {
+            ReadOnlySpan<T> span = MemoryMarshal.CreateSpan(ref input._values[0], (int)input._linearLength);
```

Comment:
What guarantees in these situations that _linearLength is <= int.MaxValue?

Reply:
Because for now the backing for Tensors is a managed array. So currently we are safe, but that will change in the future and will need to be handled better when we are dealing with native backing.